### ISSUE TITLE
TrackException doesn't fire at all.

### ISSUE THREAD

**Issue by Simnico99** [angelobreuer/Lavalink4NET#86]
**Describe the bug**

**To Reproduce**
1. Make a bot
2. Try to play this song: https://www.youtube.com/watch?v=ptzzU7jFQwo
3. IAudioService sometimes fire under TrackEnd and doesn't fire TrackException.

**Expected behavior**
IAudioService  should fire a TrackException.

**Screenshots**
![image](https://user-images.githubusercontent.com/80013536/149637355-71e7be62-2cc9-4a9c-ab12-a2090ac5ecbd.png)

**(please complete the following information):**
 - Lavalink4NET Version : 2.1.0
 - Exception thrown? N/A

**Comment by angelobreuer** [comment]
Hello @Simnico99, thank you for your report!

That seems awkward, did you subscribe to any of the player events? Or do you await on these? (This may cause to only dispatch the first event, and not receiving any of the following, because the node I/O is blocked).

**Comment by Simnico99** [comment]
I am running everything async 

this is surrounded by TaskAsync.Run wich is a non block Task that run in a seperated thread an return a Task.Completed on start so it doesn't block the Thread, It works for play and stop and TrackEnd.
(From my own Asyncronizer.Task nuget package.)

```csharp
SpinWait.SpinUntil(() => value.IsCancellationRequested, timeSpan);
```

I do subscribe to the player events here is the content of the file:

```csharp
public class PlayerService : IPlayerService
{
    private readonly ConcurrentDictionary<ulong?, CancellationTokenSource> _aloneDisconnectTokens;
    private readonly ConcurrentDictionary<ulong?, CancellationTokenSource> _disconnectTokens;
    private readonly IAudioService _audioService;
    private readonly ILogger _logger;
    private readonly LockHelper _lockHelper;

    public PlayerService(IAudioService audioService, ILogger<PlayerService> logger, LockHelper lockHelper)
    {
        _disconnectTokens = new ConcurrentDictionary<ulong?, CancellationTokenSource>();
        _aloneDisconnectTokens = new ConcurrentDictionary<ulong?, CancellationTokenSource>();
        _logger = logger;
        _audioService = audioService;
        _lockHelper = lockHelper;

        audioService.TrackEnd += AudioService_TrackEnd;
        audioService.TrackStuck += AudioService_TrackStuck;
        audioService.TrackStarted += AudioService_TrackStarted;
        audioService.TrackException += AudioService_TrackException;
    }

    private async Task AudioService_TrackException(object sender, TrackExceptionEventArgs eventArgs)
    {
        if (eventArgs.Player is ZirconPlayer player)
        {
            _logger.LogWarning("Track {TackTitle} threw an exception. Please check Lavalink console/logs.", player.CurrentTrack?.Title);

            var embed = EmbedHandler.Create(player.Context);
            embed.AddField("Error", $"Cannot play:\n{player.CurrentTrack?.Title} because of an error.");
            await player.Context.ReplyToCommandAsync(embed: embed.BuildSync(ZirconEmbedType.Error));

            if (player.Queue.Count == 0)
            {
                await player.StopAsync();
                await InitiateDisconnectAsync(player, TimeSpan.FromSeconds(40));
            }
            else
            {
                await player.SkipAsync();
            }
        }
    }

    private async Task AudioService_TrackStuck(object sender, TrackStuckEventArgs eventArgs)
    {

        if (eventArgs.Player is ZirconPlayer player)
        {

            _logger.LogWarning("Track {TrackTitle} got stuck. Please check Lavalink console/logs.", player.CurrentTrack?.Title);
            if (player.IsLooping)
            {
                if (player.CurrentTrack != null)
                {
                    var currentSong = eventArgs.TrackIdentifier;
                    var track = await _audioService.GetTrackAsync(currentSong, SearchMode.YouTube, true);
                    if (track != null)
                    {
                        await player.PlayAsync(track);
                    }
                    else
                    {
                        await player.StopAsync();
                        await InitiateDisconnectAsync(player, TimeSpan.FromSeconds(40));
                    }
                }
                else
                {
                    await player.StopAsync();
                    await InitiateDisconnectAsync(player, TimeSpan.FromSeconds(40));
                }
            }

            if (player.Queue.Count == 0 && !player.IsLooping)
            {
                await player.StopAsync();
                await InitiateDisconnectAsync(player, TimeSpan.FromSeconds(40));
            }
            else
            {
                await player.SkipAsync();
            }
        }
    }

    private async Task AudioService_TrackStarted(object sender, TrackStartedEventArgs eventArgs) => await CancelDisconnectAsync(eventArgs.Player);

    private async Task AudioService_TrackEnd(object sender, TrackEndEventArgs eventArgs)
    {
        _logger.LogDebug("Stop reason: {Reason}", eventArgs.Reason);
        if (eventArgs.Reason != TrackEndReason.LoadFailed)
        {
            if (eventArgs.Player is ZirconPlayer player)
            {
                if (player.IsLooping)
                {
                    await player.ReplayAsync();
                    return;
                }

                if (player.Queue.IsEmpty && eventArgs.Reason != TrackEndReason.Replaced)
                {
                    player.Queue.Clear();
                    await player.StopAsync();
                    await InitiateDisconnectAsync(eventArgs.Player, TimeSpan.FromSeconds(40));
                }
            }
        }
        else
        {
            if (eventArgs.Player is ZirconPlayer player)
            {
                if (player.ErrorRetry >= 5 && player.LastErrorTrack == player.CurrentTrack)
                {
                    var embed = EmbedHandler.Create(player.Context);
                    embed.AddField("Error", $"Cannot play:\n{player.CurrentTrack?.Title} because of an error.");
                    await player.Context.ReplyToCommandAsync(embed: embed.BuildSync(ZirconEmbedType.Error));

                    await InitiateDisconnectAsync(eventArgs.Player, TimeSpan.FromSeconds(40));
                    await player.SkipAsync();
                }
                else
                {
                    if (player.LastErrorTrack != player.CurrentTrack)
                    {
                        player.LastErrorTrack = player.CurrentTrack;
                        player.ErrorRetry = 0;
                    }

                    player.ErrorRetry++;

                    await player.PlayAsync(player.CurrentTrack);
                    await Task.Delay(TimeSpan.FromSeconds(2));
                }
            }
        }
    }

    public async Task CancelDisconnectAsync(LavalinkPlayer player)
    {
        if (player is ZirconPlayer zirconplayer)
        {
            _logger.LogDebug("Canceling Disconnect");
            if (!_disconnectTokens.TryGetValue(zirconplayer.VoiceChannelId ?? 0, out var value))
            {
                value = new CancellationTokenSource();
            }
            else if (value.IsCancellationRequested)
            {
                return;
            }

            await Task.Run(() => value.Cancel(true));
            _logger.LogDebug("Canceled Disconnect");
        }
    }



    public async Task CancelAloneDisconnectAsync(LavalinkPlayer player)
    {
        _logger.LogDebug("Canceling Alone Disconnect");
        if (player is ZirconPlayer zirconplayer)
        {
            await zirconplayer.Locker.LockAsync(async () =>
            {
                if (!_aloneDisconnectTokens.TryGetValue(player.VoiceChannelId ?? 0, out var value))
                {
                    value = new CancellationTokenSource();
                }
                else if (value.IsCancellationRequested)
                {
                    return;
                }

                await Task.Run(() => value.Cancel(true));
                _logger.LogDebug("Canceled Alone Disconnect");
            });
        }
    }

    public async Task InitiateDisconnectAsync(LavalinkPlayer player, TimeSpan timeSpan)
    {
        _logger.LogDebug("Disconnecting");
        if (!_disconnectTokens.TryGetValue(player.VoiceChannelId ?? 0, out var value))
        {
            value = new CancellationTokenSource();
            _disconnectTokens.TryAdd(player.VoiceChannelId, value);
        }
        else if (value.IsCancellationRequested)
        {
            _disconnectTokens.TryUpdate(player.VoiceChannelId, new CancellationTokenSource(), value);
            value = _disconnectTokens[player.VoiceChannelId ?? 0];
        }

        await TaskAsync.Run(async () =>
        {
            var isCancelled = SpinWait.SpinUntil(() => value.IsCancellationRequested, timeSpan);
            if (isCancelled)
            {
                return;
            }

            await player.DisconnectAsync();
            _logger.LogDebug("Disconnected");
        });
    }

    public async Task BotIsAloneAsync(LavalinkPlayer player, TimeSpan timeSpan)
    {
        _logger.LogDebug("Alone Disconnecting");
        if (!_aloneDisconnectTokens.TryGetValue(player.VoiceChannelId ?? 0, out var value))
        {
            value = new CancellationTokenSource();
            _aloneDisconnectTokens.TryAdd(player.VoiceChannelId, value);
        }
        else if (value.IsCancellationRequested)
        {
            _aloneDisconnectTokens.TryUpdate(player.VoiceChannelId, new CancellationTokenSource(), value);
            value = _aloneDisconnectTokens[player.VoiceChannelId ?? 0];
        }

        var isCancelled = SpinWait.SpinUntil(() => value.IsCancellationRequested, timeSpan);
        if (isCancelled)
        {
            return;
        }

        await player.DisconnectAsync();
        _logger.LogDebug("Alone Disconnected");
    }
}

```

**Comment by Simnico99** [comment]
P.S
ZirconPlayer is just a modified QueuedLavalinkPlayer:
```csharp
using ZirconSound.Helpers;

namespace ZirconSound.Services;

public class ZirconPlayer : QueuedLavalinkPlayer
{
    public LockHelper Locker { get; set; } = new LockHelper();
    public int ErrorRetry { get; set; }
    public LavalinkTrack LastErrorTrack { get; set; }
    public IInteractionContext Context { get; private set; }

    public void SetInteraction(IInteractionContext context) => Context = context;
}
```

**Comment by angelobreuer** [comment]
@Simnico99 

Thank you for providing your repository, it really helped much to find the bug. I was able to reproduce the bug. If the specific track ( https://www.youtube.com/watch?v=ptzzU7jFQwo) is played, it causes a track exception on the node.  The node sends it to the lavalink client (the bot). A few months ago, there was a change in the track exception payload, as far as I remember, and the new behavior is not handled correctly. The client expects a property named 'exception' on the JSON payload that informs the client about the exception, but the property was removed. 

I will release a fix for this problem in a few minutes or hours. It may take a while to publish the packages to NuGet as the upload tokens expired this year.
